core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size().toInt) {
```

Comment:
Thanks for the explanation. The changes look good to me then. Note that originalSize is always <= channel.size. So, it seems that we don't need to further check targetSize > channel.size since it's implied by targetSize > originalSize already.